### PR TITLE
OBMIN-344 new service for receiving the validation code

### PR DESCRIPTION
**Pull request by Sershok**
Created new service for receiving the validation code

1.  created methods **validateEmail()** in MailService
2.  add new column **validatedEmail** in User 
3.  created unit test **EmailServiceTest**

### CODE REVIEW

File: src/main/java/space/obminyashka/items_exchange/service/MailService.java [hunk 1]
```diff
@@ -1,18 +1,32 @@
...
 
     /**
      * Send a simple email to an end-user with provided locale
-     * @param emailTo a receiver email address
+     *
+     * @param emailTo   a receiver email address
      * @param emailType one of the supported email types to be sent
-     * @param locale user locale
+     * @param locale    user locale
      * @throws IOException when service is unavailable or some unexpected case happened
```
Comment (IllyaZuew) [inline src/main/java/space/obminyashka/items_exchange/service/MailService.java:19]: The same problem with spaces

File: src/main/java/space/obminyashka/items_exchange/service/impl/SendGridService.java [hunk 2]
```diff
@@ -35,6 +45,19 @@ public void sendMail(String emailTo, EmailType subject, Locale locale) throws IO
...
+    @Override
+    public void validateEmail(UUID validationCode) throws EmailValidationCodeNotFoundException, EmailValidationCodeExpiredException {
+        Optional<EmailConfirmationToken> optionalEmail = emailConfirmationTokenRepository.findById(validationCode);
+
+        if (optionalEmail.isEmpty()){
+            throw new EmailValidationCodeNotFoundException("Email not exists");
+        }
+        if(optionalEmail.get().getExpiryDate().isBefore(LocalDateTime.now())){
+            throw new EmailValidationCodeExpiredException("Validation code expired");
+        }
```
Comment (IllyaZuew) [inline src/main/java/space/obminyashka/items_exchange/service/impl/SendGridService.java:57]: will be good to rewrite it without using if but it only my opinion

File: src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    @Mock
+    private EmailConfirmationTokenRepository emailConfirmationTokenRepository;
+    @Captor
+    private ArgumentCaptor<String> userArgumentCaptor;
+    @InjectMocks
+    private SendGridService sendGridService;
+
+    @Test
+    void TestValidateEmailCode_WhenDataCorrect_Successfully() throws EmailValidationCodeExpiredException, EmailValidationCodeNotFoundException {
+        when(emailConfirmationTokenRepository.findById(UUID.fromString("61731cc8-8104-49f0-b2c3-5a52e576ab18")))
```
Comment (IllyaZuew) [inline src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java:41]: you should create variable for UUID and use it

File: src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    private Optional<EmailConfirmationToken> creatEmailConfirmToken(){
+        EmailConfirmationToken emailConfirmationToken = new EmailConfirmationToken();
+
+        LocalDateTime localDateTime = LocalDateTime.now();
+        emailConfirmationToken.setExpiryDate(localDateTime.plusDays(1L));
+
+        User user = new User();
+        user.setId(UUID.fromString("61731cc8-8104-49f0-b2c3-5a52e576ab18"));
```
Comment (IllyaZuew) [inline src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java:56]: you should create variable for UUID and use it like string

File: src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        emailConfirmationToken.setExpiryDate(localDateTime.plusDays(1L));
+
+        User user = new User();
+        user.setId(id);
+        user.setEmail("user@mail.ua");
+        emailConfirmationToken.setUser(user);
+
+        return Optional.of(emailConfirmationToken);
+    }
+}
```
Comment (IllyaZuew) [inline src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java:61]: you should write tests for exceptions and maybe you can do it like parameterized test

Comment (IllyaZuew) [reply src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java:61]: you check only successful result but you need to check and bad results with exceptions

File: src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    void TestValidateEmailCode_WhenDataCorrect_Successfully() throws EmailValidationCodeExpiredException, EmailValidationCodeNotFoundException {
+        when(emailConfirmationTokenRepository.findById(id)).thenReturn(creatEmailConfirmToken());
+        sendGridService.validateEmail(id);
+
+        verify(userRepository).setValidatedEmailToUserById(IdArgumentCaptor.capture());
+        assertEquals(id, IdArgumentCaptor.getValue());
+    }
+
+    @Test
+    void shouldThrowEmailValidationCodeNotFoundException() {
```
Comment (IllyaZuew) [inline src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java:56]: make name with patern method_when_should

File: src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        when(emailConfirmationTokenRepository.findById(id)).thenReturn(Optional.empty());
+
+        assertThrows(EmailValidationCodeNotFoundException.class, () -> sendGridService.validateEmail(id));
+    }
+
+    @Test
+    void shouldThrowEmailValidationCodeExpiredException() {
```
Comment (IllyaZuew) [inline src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java:63]: make name with patern method_when_should

**Review by IllyaZuew — APPROVED** [review]

**Review by IllyaZuew — APPROVED** [review]

**Review by ennnbaev — APPROVED** [review]

**Review by demus-n — DISMISSED** [review]

**Review by ImAndVlad — APPROVED** [review]

File: src/main/java/space/obminyashka/items_exchange/dao/UserRepository.java [hunk 8]
```diff
@@ -52,4 +52,9 @@ public interface UserRepository extends JpaRepository<User, UUID> {
     @Modifying
     @Query("update User u set u.oauth2Login = true where u.email = :email and u.oauth2Login is null")
     void setOAuth2LoginToUserByEmail(String email);
+
+    @Transactional
+    @Modifying
+    @Query("update User u set u.validatedEmail = true where u.id = :id")
+    void setValidatedEmailToUserById(UUID id);
```
Comment (rpkyrych) [inline src/main/java/space/obminyashka/items_exchange/dao/UserRepository.java:59]: We shouldn't expose users' ID outside. That method have to work with EmailConfirmationCode ID instead

File: src/main/java/space/obminyashka/items_exchange/service/impl/SendGridService.java [hunk 9]
```diff
@@ -32,9 +35,11 @@
 @Slf4j
 public class SendGridService implements MailService {
 
-    private final EmailConfirmationTokenRepository emailRepository;
+    private final EmailConfirmationCodeRepository emailRepository;
     private final SendGrid sendGrid;
     private final Email sender;
+    private final UserRepository userRepository;
+    private final EmailConfirmationCodeRepository emailConfirmationCodeRepository;
```
Comment (rpkyrych) [inline src/main/java/space/obminyashka/items_exchange/service/impl/SendGridService.java:42]: The same class is introduced on line 38

File: src/main/resources/liquibase/db.changelog-1.2.xml [hunk 10]
```diff
@@ -50,4 +50,20 @@
...
+
+    <changeSet id="1.2.4: rename table" author="serhii.ponomarenko">
+        <renameTable oldTableName="email_confirmation_token" newTableName="email_confirmation_code"/>
+    </changeSet>
+
+    <changeSet id="1.2.5: remove column token" author="serhii.ponomarenko">
+        <dropColumn tableName="email_confirmation_code">
+            <column name="token"/>
+        </dropColumn>
+    </changeSet>
```
Comment (rpkyrych) [inline src/main/resources/liquibase/db.changelog-1.2.xml:68]: Don't you mind making all those changes by a single changeset?

**Review by rpkyrych — CHANGES_REQUESTED** [review]
Some parts need to be changed.

**Pull request closed without merging**